### PR TITLE
[common] Improve symbolic polynomial performance via is_expanded

### PR DESCRIPTION
**Pull request by jwnimmer-tri**
Towards #16292.

In cases where we are operating on affine expressions over variables, notice when `is_expanded == true` must be true during construction, so that we don't need to re-expand it later.

This yields a ~2x speedup in `BenchmarkSosProgram1` (from approximately 9.8 seconds runtime to 4.3 seconds runtime, on my machine).

As of this change, all calls from the benchmark program to `Expression::Expand` are no-ops; the expression is always already known to be expanded.  Specifically, `bazel run //solvers/benchmarking:benchmark_mathematical_program` passes when using this patch to fail-fast on non-noops:

```patch
diff --git a/common/symbolic_expression.cc b/common/symbolic_expression.cc
index db191a6d0e..d12a617084 100644
--- a/common/symbolic_expression.cc
+++ b/common/symbolic_expression.cc
@@ -213,7 +213,7 @@ Expression Expression::Expand() const {
     // Expand() on the cell.
     return *this;
   }
-
+  DRAKE_UNREACHABLE();
   Expression result = cell().Expand();
   if (!result.is_expanded()) {
     result.mutable_cell().set_expanded();
```

---
This change is [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/16980)

### CODE REVIEW

**Comment by jwnimmer-tri** [comment]
+@hongkai-dai for feature review, please.

In particular, please double-check the soundness of my reasoning for when an ExpressionAdd or ExpressionMul cell can know `is_expanded() == true` at construction-time (i.e., each factory's logic for `is_expanded_`), or if there are any corner cases that I've missed.

Getting that reasoning correct was the most difficult part of this change. I did my best to probe it with unit test tests and write clear comments explaining the thought process, but I could have missed something.

**Review by hongkai-dai — COMMENTED** [review]
@jwnimmer-tri 

First of all, thanks a lot for working on this. The performance improvement is great.

Here I try to summarize my reasoning/understanding of the PR:
1. The benchmark_mathematical_program.cc file shows that the speed bottleneck is in computing the product and addition of polynomials. This slowness is caused when we compute the product and sum of expressions (where the expression being the coefficients of polynomials) we keep calling Expand() in every step of DoAddProduct() function. So the goal is to remove the redundant call of Expand function. If an expression is already expanded (Expression::is_expanded() returns true), then we can bypass calling `Expand()` function. So we want to set `is_expanded()` to true as much as we can, especially when we construct the expression in the first place.
2. The coefficients of the polynomials are either ExpressionConstant or ExpressionVar, their constructors already set `is_expanded` to true.
3. We then use ExpressionAddFactory and ExpressionMulFactory to produce ExpressionAdd and ExpressionMul. Hence we should modify these two classes to set `is_expanded_` flag whenever we can.

Is this the right understanding?

I reviewed the modification on ExpressionAddFactory, but have some questions, could you let me know what I miss? Thanks!

> Reviewed 1 of 3 files at r1.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/16980#-:-N-oZ5xj1reYwB_ubGkh:buhdhb1)* status: 3 unresolved discussions, LGTM missing from assignee hongkai-dai, needs platform reviewer assigned, needs at least two assigned reviewers (waiting on @hongkai-dai and @jwnimmer-tri)

___
*[common/symbolic_expression_cell.cc, line 609 at r1](https://reviewable.io/reviews/robotlocomotion/drake/16980#-N-ocfFI1u5vA8EjQrlW:-N-ocfFI1u5vA8EjQrlX:b-eh03hg) ([raw file](https://github.com/robotlocomotion/drake/blob/2118baf77f3e39b354a59fa4f36e45cb53bb77b5/common/symbolic_expression_cell.cc#L609)):*
> ```cpp
> ExpressionAddFactory::ExpressionAddFactory(
>     const double constant, map<Expression, double> expr_to_coeff_map)
>     : is_expanded_(false), constant_{constant},
> ```

If each Expression in the key of expr_to_coeff_map is expanded, then `is_expanded_` should be true? Are we being conservative here?
___
*[common/symbolic_expression_cell.cc, line 696 at r1](https://reviewable.io/reviews/robotlocomotion/drake/16980#-N-odgY81_6JE2jHXDU9:-N-odgY81_6JE2jHXDUA:b-g7cdds) ([raw file](https://github.com/robotlocomotion/drake/blob/2118baf77f3e39b354a59fa4f36e45cb53bb77b5/common/symbolic_expression_cell.cc#L696)):*
> ```cpp
>       // TODO(soonho-tri): The following operation is not sound since it cancels
>       // `term` which might contain 0/0 problems.
>       expr_to_coeff_map_.erase(it);
> ```

Should we also check that after erasing `it`, the remaining keys in expr_to_coeff_map_ are all expanded? If yes then we can set is_expanded_ to true?
___
*[common/symbolic_expression_cell.cc, line 705 at r1](https://reviewable.io/reviews/robotlocomotion/drake/16980#-N-oe3-TE9ePxRMALDnM:-N-oe3-TE9ePxRMALDnN:b-kpvnkf) ([raw file](https://github.com/robotlocomotion/drake/blob/2118baf77f3e39b354a59fa4f36e45cb53bb77b5/common/symbolic_expression_cell.cc#L705)):*
> ```cpp
>     // another Expression inside of itself), then conservatively we can no
>     // longer be sure that our ExpressionAdd remains in expanded form.
>     if (!is_variable(term)) {
> ```

Not sure if I follow this logic completely. If term.is_expanded() is true, then I think we don't need to modify `is_expanded_`? Namely shall we add the condition as
```cc
if (!term.is_expanded() && !is_variable(term)) {
  is_expanded_ = false;
}
```

**Review by jwnimmer-tri — COMMENTED** [review]
> The benchmark_mathematical_program.cc file shows that the speed bottleneck is in computing the product and addition of polynomials.

Yes, that's true.

> This slowness is caused when we compute the product and sum of expressions (where the expression being the coefficients of polynomials) we keep calling Expand() in every step of DoAddProduct() function.

That's true in part.  There are a few other places (besides `Expand()`) where we are doing useless operations.  Those will in follow-up PRs.

> So the goal is to remove the redundant call of Expand function. If an expression is already expanded (Expression::is_expanded() returns true), then we can bypass calling Expand() function. So we want to set is_expanded() to true as much as we can, especially when we construct the expression in the first place.

Yes, exactly.

> The coefficients of the polynomials are either ExpressionConstant or ExpressionVar, their constructors already set is_expanded to true.

Yes.

> We then use ExpressionAddFactory and ExpressionMulFactory to produce ExpressionAdd and ExpressionMul. Hence we should modify these two classes to set is_expanded_ flag whenever we can.

Yes.  The key observation is that in our use case, the coeffs are always affine (and the monomials are always just products or powers of vars).  Thus, the Expression codepaths necessary to create and operate on them are very narrow, and it's not too difficult to add a few very inexpensive checks, for a large payoff.

Any optimizations beyond those necessary for affine coeffs are deferred as future work.

> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/16980#-:-N-wzSVa5oPgccVwKNv-:b-t02x0f)* status: 3 unresolved discussions, LGTM missing from assignee hongkai-dai, needs platform reviewer assigned, needs at least two assigned reviewers (waiting on @hongkai-dai and @jwnimmer-tri)

___
*[common/symbolic_expression_cell.cc, line 609 at r1](https://reviewable.io/reviews/robotlocomotion/drake/16980#-N-ocfFI1u5vA8EjQrlW:-N-wv0v3AK3chbUzl01u:b-r6ceig) ([raw file](https://github.com/robotlocomotion/drake/blob/2118baf77f3e39b354a59fa4f36e45cb53bb77b5/common/symbolic_expression_cell.cc#L609)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

If each Expression in the key of expr_to_coeff_map is expanded, then `is_expanded_` should be true? Are we being conservative here?
</blockquote></details>

Yes.  I'll work on adding a code comment here to clarify.

I used to say `is_expanded_(expr_to_coeff_map.empty())` here which is pretty clearly true, but there was no way to unit test it via the `Expression` API.  I think the same would probably occur for checking the map keys as `is_exapanded()`.

In practice, any map we pass into here seems to always have non-expanded terms.  There are a lot of short-circuits in the Expression operators (e.g., [here](https://github.com/RobotLocomotion/drake/blob/a74e38e04c76b91650833f6292c8363fcde4a965/common/symbolic_expression.cc#L448-L453)) that end up curtailing what is passed into here in practice.

Also, checking the map `is_expanded()` in a loop might not be a fast operation (depending on the map size), so it's not obvious that doing so would result in a net performance gain.

In short: I don't yet think there's any benefit to doing something more complicated here, and there's definitely a complexity cost (and runtime cost) to adding more logic here.
___
*[common/symbolic_expression_cell.cc, line 696 at r1](https://reviewable.io/reviews/robotlocomotion/drake/16980#-N-odgY81_6JE2jHXDU9:-N-wvy69BVsquULdsq0K:bw903yg) ([raw file](https://github.com/robotlocomotion/drake/blob/2118baf77f3e39b354a59fa4f36e45cb53bb77b5/common/symbolic_expression_cell.cc#L696)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

Should we also check that after erasing `it`, the remaining keys in expr_to_coeff_map_ are all expanded? If yes then we can set is_expanded_ to true?
</blockquote></details>

Checking the map `is_expanded()` in a loop might not be a fast operation (depending on the map size), so it's not obvious that doing so would result in a net performance gain.  Since this PR gives us `is_expanded() == true` for the cases we care about in the benchmark, I am satisfied to leave the code here unchanged, i.e., being conservative.
___
*[common/symbolic_expression_cell.cc, line 705 at r1](https://reviewable.io/reviews/robotlocomotion/drake/16980#-N-oe3-TE9ePxRMALDnM:-N-wxSv0EAi28r8WdklA:bi2eisl) ([raw file](https://github.com/robotlocomotion/drake/blob/2118baf77f3e39b354a59fa4f36e45cb53bb77b5/common/symbolic_expression_cell.cc#L705)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

Not sure if I follow this logic completely. If term.is_expanded() is true, then I think we don't need to modify `is_expanded_`? Namely shall we add the condition as
```cc
if (!term.is_expanded() && !is_variable(term)) {
  is_expanded_ = false;
}
```
</blockquote></details>

(Aside: note that an ExpressionVar is always `is_expanded() == true`, so only the first half of the predicate is necessary / relevant.)

If we make that change, the unit test case for `5 * (3 + 2y) + 30 * (7 + x_)` fails.

I'll work on figuring out why that's unsound, so then I can write it here into a comment.

**Review by jwnimmer-tri — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/16980)* status: 3 unresolved discussions, LGTM missing from assignee hongkai-dai, needs platform reviewer assigned, needs at least two assigned reviewers (waiting on @hongkai-dai and @jwnimmer-tri)

___
*[common/symbolic_expression_cell.cc, line 705 at r1](https://reviewable.io/reviews/robotlocomotion/drake/16980#-N-oe3-TE9ePxRMALDnM:-N-xR4bH9jTsWba8PP8r:bkcpy0u) ([raw file](https://github.com/robotlocomotion/drake/blob/2118baf77f3e39b354a59fa4f36e45cb53bb77b5/common/symbolic_expression_cell.cc#L705)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

(Aside: note that an ExpressionVar is always `is_expanded() == true`, so only the first half of the predicate is necessary / relevant.)

If we make that change, the unit test case for `5 * (3 + 2y) + 30 * (7 + x_)` fails.

I'll work on figuring out why that's unsound, so then I can write it here into a comment.
</blockquote></details>

The term `5 * (3 + 2 * y)` ends up calling `ExpressionAddFactory::AddTerm(5.0, (3 + 2 * y))`.  Even though the `3 + 2 * y` has `is_expanded() == true`, it's not the case that the AddFactory remains expanded -- indeed, without distributing the `5` across the nested addition, we are no longer expanded.  I'll try to work this into a comment.

**Review by hongkai-dai — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/16980)* status: 2 unresolved discussions, LGTM missing from assignee hongkai-dai, needs platform reviewer assigned, needs at least two assigned reviewers (waiting on @hongkai-dai and @jwnimmer-tri)

___
*[common/symbolic_expression_cell.cc, line 609 at r1](https://reviewable.io/reviews/robotlocomotion/drake/16980#-N-ocfFI1u5vA8EjQrlW:-N-xRC9N5TrayBtOrT8N:bhxwavz) ([raw file](https://github.com/robotlocomotion/drake/blob/2118baf77f3e39b354a59fa4f36e45cb53bb77b5/common/symbolic_expression_cell.cc#L609)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Yes.  I'll work on adding a code comment here to clarify.

I used to say `is_expanded_(expr_to_coeff_map.empty())` here which is pretty clearly true, but there was no way to unit test it via the `Expression` API.  I think the same would probably occur for checking the map keys as `is_exapanded()`.

In practice, any map we pass into here seems to always have non-expanded terms.  There are a lot of short-circuits in the Expression operators (e.g., [here](https://github.com/RobotLocomotion/drake/blob/a74e38e04c76b91650833f6292c8363fcde4a965/common/symbolic_expression.cc#L448-L453)) that end up curtailing what is passed into here in practice.

Also, checking the map `is_expanded()` in a loop might not be a fast operation (depending on the map size), so it's not obvious that doing so would result in a net performance gain.

In short: I don't yet think there's any benefit to doing something more complicated here, and there's definitely a complexity cost (and runtime cost) to adding more logic here.
</blockquote></details>

That makes sense, thanks for the explanation.
___
*[common/symbolic_expression_cell.cc, line 696 at r1](https://reviewable.io/reviews/robotlocomotion/drake/16980#-N-odgY81_6JE2jHXDU9:-N-xRHwsCUisIu4k2C33:b1yptmk) ([raw file](https://github.com/robotlocomotion/drake/blob/2118baf77f3e39b354a59fa4f36e45cb53bb77b5/common/symbolic_expression_cell.cc#L696)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Checking the map `is_expanded()` in a loop might not be a fast operation (depending on the map size), so it's not obvious that doing so would result in a net performance gain.  Since this PR gives us `is_expanded() == true` for the cases we care about in the benchmark, I am satisfied to leave the code here unchanged, i.e., being conservative.
</blockquote></details>

Got it, that makes sense.
___
*[common/symbolic_expression_cell.cc, line 705 at r1](https://reviewable.io/reviews/robotlocomotion/drake/16980#-N-oe3-TE9ePxRMALDnM:-N-xSH0j6z8KxtfMom2q:bsds6nd) ([raw file](https://github.com/robotlocomotion/drake/blob/2118baf77f3e39b354a59fa4f36e45cb53bb77b5/common/symbolic_expression_cell.cc#L705)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

The term `5 * (3 + 2 * y)` ends up calling `ExpressionAddFactory::AddTerm(5.0, (3 + 2 * y))`.  Even though the `3 + 2 * y` has `is_expanded() == true`, it's not the case that the AddFactory remains expanded -- indeed, without distributing the `5` across the nested addition, we are no longer expanded.  I'll try to work this into a comment.
</blockquote></details>

I see.

I think this reminds me another use case in sum-of-squares optimization. Currently in `benchmark_mathematical_program`, we have `p_i(x).dot(q_i(x))`, where the coefficients of `p_i(x)` are just doubles, and the coefficients of `q_i(x)` are just symbolic variables. I can have another case that the coefficients of `p_i(x)` are still doubles, but the coefficients of `q_i(x)` are affine expressions. In that case, I think we won't be able to detect that the product of their expression is expanded.

Sorry I missed this case when I created `benchmark_mathematical_program`, I will add the case in a separate PR.

**Review by jwnimmer-tri — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/16980)* status: 1 unresolved discussion, LGTM missing from assignee hongkai-dai, needs platform reviewer assigned, needs at least two assigned reviewers (waiting on @hongkai-dai)

___
*[common/symbolic_expression_cell.cc, line 705 at r1](https://reviewable.io/reviews/robotlocomotion/drake/16980#-N-oe3-TE9ePxRMALDnM:-N-xTJ7x8HO9VXBHUSkA:b-4vtlv9) ([raw file](https://github.com/robotlocomotion/drake/blob/2118baf77f3e39b354a59fa4f36e45cb53bb77b5/common/symbolic_expression_cell.cc#L705)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

I see.

I think this reminds me another use case in sum-of-squares optimization. Currently in `benchmark_mathematical_program`, we have `p_i(x).dot(q_i(x))`, where the coefficients of `p_i(x)` are just doubles, and the coefficients of `q_i(x)` are just symbolic variables. I can have another case that the coefficients of `p_i(x)` are still doubles, but the coefficients of `q_i(x)` are affine expressions. In that case, I think we won't be able to detect that the product of their expression is expanded.

Sorry I missed this case when I created `benchmark_mathematical_program`, I will add the case in a separate PR.
</blockquote></details>

There is a lot of short-circuiting already happening in other places, so there's some chance it might already work optimally.  In any case, I can revisit once we have the updated benchmark.  For now, I've added comments in this PR to explain what we have so far.

**Review by hongkai-dai — COMMENTED** [review]
:lgtm_strong: +@sherm1 for platform review please, thanks!

> Reviewed 1 of 3 files at r1, 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/16980#-:-N-xmFZG5jZh6PqKqG9L:b37312u)* status: 1 unresolved discussion, LGTM missing from assignee sherm1(platform) (waiting on @jwnimmer-tri and @sherm1)

___
*[common/symbolic_expression_cell.cc, line 705 at r1](https://reviewable.io/reviews/robotlocomotion/drake/16980#-N-oe3-TE9ePxRMALDnM:-N-xU6JGAA2_LKu3RbWf:bil7lsi) ([raw file](https://github.com/robotlocomotion/drake/blob/2118baf77f3e39b354a59fa4f36e45cb53bb77b5/common/symbolic_expression_cell.cc#L705)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

There is a lot of short-circuiting already happening in other places, so there's some chance it might already work optimally.  In any case, I can revisit once we have the updated benchmark.  For now, I've added comments in this PR to explain what we have so far.
</blockquote></details>

Sounds good. Thanks!
___
*[common/symbolic_expression_cell.cc, line 1044 at r2](https://reviewable.io/reviews/robotlocomotion/drake/16980#-N-xl5Yj-_8hZzJ6lKJo:-N-xl5YkBzPRJ8MjjCZr:b-jd49su) ([raw file](https://github.com/robotlocomotion/drake/blob/248c29a5351c27fc04d029f7d935c20b0cf84b8c/common/symbolic_expression_cell.cc#L1044)):*
> ```cpp
>     // that our ExpressionMul remains in expanded form.
>     if (!((is_variable(base) && is_constant(exponent)) ||
>           (is_constant(base) && is_variable(exponent)))) {
> ```

Maybe also worth mentioning that power(variable, variable) is not detected with is_expanded()=True yet, so we don't include it here?

**Review by jwnimmer-tri — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/16980)* status: 1 unresolved discussion, LGTM missing from assignee sherm1(platform) (waiting on @jwnimmer-tri and @sherm1)

___
*[common/symbolic_expression_cell.cc, line 1044 at r2](https://reviewable.io/reviews/robotlocomotion/drake/16980#-N-xl5Yj-_8hZzJ6lKJo:-N-xo1Ja2hq_Z2Op0GWf:bc6c7f) ([raw file](https://github.com/robotlocomotion/drake/blob/248c29a5351c27fc04d029f7d935c20b0cf84b8c/common/symbolic_expression_cell.cc#L1044)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

Maybe also worth mentioning that power(variable, variable) is not detected with is_expanded()=True yet, so we don't include it here?
</blockquote></details>

That's interesting.  It might be both clearer and more performant to recognize `exp(var, var)` as pre-expanded.  I will play with it locally to see what makes the most sense.

**Review by sherm1 — COMMENTED** [review]
Platform :lgtm: with a few comments

> Reviewed 2 of 3 files at r1, 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/16980#-:-N-xz4Mx0lSQGkDI0q3E:b-tv2fmf)* status: 2 unresolved discussions (waiting on @jwnimmer-tri)

___
*[common/symbolic_expression_cell.h, line 320 at r2](https://reviewable.io/reviews/robotlocomotion/drake/16980#-N-xv8X31cC5sx6MoK-J:-N-xv8X31cC5sx6MoK-K:b9vdj06) ([raw file](https://github.com/robotlocomotion/drake/blob/248c29a5351c27fc04d029f7d935c20b0cf84b8c/common/symbolic_expression_cell.h#L320)):*
> ```objc
>   void AddMap(const std::map<Expression, double>& expr_to_coeff_map);
> 
>   bool is_expanded_{true};
> ```

BTW it worries me to see the default here as `true`. Why not default to the more conservative choice? Then a constructor that can guarantee expansion can explicitly set it to `true` and everyone else can do nothing.
___
*[common/symbolic_expression_cell.cc, line 619 at r2](https://reviewable.io/reviews/robotlocomotion/drake/16980#-N-xvzPyANdjROoSk7o9:-N-xvzPyANdjROoSk7oA:b-w49n6a) ([raw file](https://github.com/robotlocomotion/drake/blob/248c29a5351c27fc04d029f7d935c20b0cf84b8c/common/symbolic_expression_cell.cc#L619)):*
> ```cpp
>     const ExpressionAdd& add)
>     : ExpressionAddFactory{add.get_constant(), add.get_expr_to_coeff_map()} {
>   is_expanded_ = add.is_expanded();
> ```

BTW could set the member in the initialization list
___
*[common/test/symbolic_expansion_test.cc, line 87 at r2](https://reviewable.io/reviews/robotlocomotion/drake/16980#-N-xydsi3uYxQ521jS5J:-N-xydsi3uYxQ521jS5K:bkqimbw) ([raw file](https://github.com/robotlocomotion/drake/blob/248c29a5351c27fc04d029f7d935c20b0cf84b8c/common/test/symbolic_expansion_test.cc#L87)):*
> ```cpp
>   EXPECT_TRUE(CheckAlreadyExpanded(-x_));
>   EXPECT_TRUE(CheckAlreadyExpanded(3 * x_));
>   EXPECT_TRUE(CheckAlreadyExpanded(-(2 * x_)));
> ```

minor: this replaced the test "-2 * x_". Shouldn't that also remain expanded (with constant=-2)? Why remove that test?

**Review by jwnimmer-tri — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/16980)* status: 1 unresolved discussion (waiting on @hongkai-dai and @sherm1)

___
*[common/symbolic_expression_cell.h, line 320 at r2](https://reviewable.io/reviews/robotlocomotion/drake/16980#-N-xv8X31cC5sx6MoK-J:-N-y0dpV4vGwXiPMsWQ0:bg5jxkj) ([raw file](https://github.com/robotlocomotion/drake/blob/248c29a5351c27fc04d029f7d935c20b0cf84b8c/common/symbolic_expression_cell.h#L320)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW it worries me to see the default here as `true`. Why not default to the more conservative choice? Then a constructor that can guarantee expansion can explicitly set it to `true` and everyone else can do nothing.
</blockquote></details>

OK This is correct value for the default constructor (i.e., of an empty factory).  The condition starts as being "trivially true", until proven otherwise.  If this said `false`, then a lot of the tests would fail.  I could replace `ExpressionAddFactory() = default;` with an out-of-line constructor that provided for the "true" in thata case, but I think that's more confusing on balance.  There are enough unit tests of this subsystem that any bugs related to this field will be caught quickly.  Mostly it's the mutation operations that make us false, not any constructor things.
___
*[common/symbolic_expression_cell.cc, line 619 at r2](https://reviewable.io/reviews/robotlocomotion/drake/16980#-N-xvzPyANdjROoSk7o9:-N-y1II-90rOfraHQMdj:b-2nymow) ([raw file](https://github.com/robotlocomotion/drake/blob/248c29a5351c27fc04d029f7d935c20b0cf84b8c/common/symbolic_expression_cell.cc#L619)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW could set the member in the initialization list
</blockquote></details>

I don't think that's true.  This patch

```patch
--- a/common/symbolic_expression_cell.cc
+++ b/common/symbolic_expression_cell.cc
@@ -615,9 +615,8 @@ ExpressionAddFactory::ExpressionAddFactory(
 
 ExpressionAddFactory::ExpressionAddFactory(
     const ExpressionAdd& add)
-    : ExpressionAddFactory{add.get_constant(), add.get_expr_to_coeff_map()} {
-  is_expanded_ = add.is_expanded();
-}
+    : ExpressionAddFactory{add.get_constant(), add.get_expr_to_coeff_map()},
+      is_expanded_{add.is_expanded()} {}
 
 void ExpressionAddFactory::AddExpression(const Expression& e) {
   if (is_constant(e)) {
```

produces

```
common/symbolic_expression_cell.cc:619:37: error: mem-initializer for 'drake::symbolic::ExpressionAddFactory::is_expanded_' follows constructor delegation
  619 |       is_expanded_{add.is_expanded()} {}
```
___
*[common/symbolic_expression_cell.cc, line 1044 at r2](https://reviewable.io/reviews/robotlocomotion/drake/16980#-N-xl5Yj-_8hZzJ6lKJo:-N-y29M92mwYcSpqYdAp:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/248c29a5351c27fc04d029f7d935c20b0cf84b8c/common/symbolic_expression_cell.cc#L1044)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

That's interesting.  It might be both clearer and more performant to recognize `exp(var, var)` as pre-expanded.  I will play with it locally to see what makes the most sense.
</blockquote></details>

Done.
___
*[common/test/symbolic_expansion_test.cc, line 87 at r2](https://reviewable.io/reviews/robotlocomotion/drake/16980#-N-xydsi3uYxQ521jS5J:-N-y1foEEofOaoZYPG3z:b-tljttz) ([raw file](https://github.com/robotlocomotion/drake/blob/248c29a5351c27fc04d029f7d935c20b0cf84b8c/common/test/symbolic_expansion_test.cc#L87)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

minor: this replaced the test "-2 * x_". Shouldn't that also remain expanded (with constant=-2)? Why remove that test?
</blockquote></details>

OK sure, no harm in keeping it, though it's an exact functional duplicate of the `3 * x_` above.  (The constant `-2` vs `3` makes no difference.)

**Review by hongkai-dai — COMMENTED** [review]
> Reviewed 2 of 2 files at r3.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/16980)* status: 1 unresolved discussion (waiting on @jwnimmer-tri)

___
*[common/test/symbolic_expansion_test.cc, line 104 at r3](https://reviewable.io/reviews/robotlocomotion/drake/16980#-N-y2u_dDT37NX8hSb38:-N-y2u_dDT37NX8hSb39:b-kempgo) ([raw file](https://github.com/robotlocomotion/drake/blob/327c82bfd5c2103dbfcd0abec77c7e681a5eb57a/common/test/symbolic_expansion_test.cc#L104)):*
> ```cpp
>   // `is_expanded() == true` upon construction, but in any case they must not
>   // change form when `Expand()` is called.
>   EXPECT_TRUE(CheckUnchangedExpand(pow(x_, y_)));            // x^y
> ```

I think we should recognize is_expanded() == true for pow(x_, y_), based on the most recent commit?

**Review by jwnimmer-tri — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/16980)* status: 1 unresolved discussion (waiting on @hongkai-dai)

___
*[common/test/symbolic_expansion_test.cc, line 104 at r3](https://reviewable.io/reviews/robotlocomotion/drake/16980#-N-y2u_dDT37NX8hSb38:-N-y54P77_3yPfPNkoo2:blipx9i) ([raw file](https://github.com/robotlocomotion/drake/blob/327c82bfd5c2103dbfcd0abec77c7e681a5eb57a/common/test/symbolic_expansion_test.cc#L104)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

I think we should recognize is_expanded() == true for pow(x_, y_), based on the most recent commit?
</blockquote></details>

Done.

This one doesn't pass yet (it never uses the ExpressionMulFactory), but you're right that there was a test missing.

I added a new test case above, and confirmed that it failed against r2 but passes now with r3.

**Review by hongkai-dai — COMMENTED** [review]
> Reviewed 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/16980)* status: :shipit: complete! all discussions resolved, LGTM from assignees sherm1(platform),hongkai-dai (waiting on @jwnimmer-tri)

**Comment by sherm1** [comment]
*[common/symbolic_expression_cell.cc, line 619 at r2](https://reviewable.io/reviews/robotlocomotion/drake/16980#-N-xvzPyANdjROoSk7o9:-N-yOjRYD49lcyTEwCWo:b8j6veb) ([raw file](https://github.com/robotlocomotion/drake/blob/248c29a5351c27fc04d029f7d935c20b0cf84b8c/common/symbolic_expression_cell.cc#L619)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

I don't think that's true.  This patch

```patch
--- a/common/symbolic_expression_cell.cc
+++ b/common/symbolic_expression_cell.cc
@@ -615,9 +615,8 @@ ExpressionAddFactory::ExpressionAddFactory(
 
 ExpressionAddFactory::ExpressionAddFactory(
     const ExpressionAdd& add)
-    : ExpressionAddFactory{add.get_constant(), add.get_expr_to_coeff_map()} {
-  is_expanded_ = add.is_expanded();
-}
+    : ExpressionAddFactory{add.get_constant(), add.get_expr_to_coeff_map()},
+      is_expanded_{add.is_expanded()} {}
 
 void ExpressionAddFactory::AddExpression(const Expression& e) {
   if (is_constant(e)) {
```

produces

```
common/symbolic_expression_cell.cc:619:37: error: mem-initializer for 'drake::symbolic::ExpressionAddFactory::is_expanded_' follows constructor delegation
  619 |       is_expanded_{add.is_expanded()} {}
```
</blockquote></details>

Oh, TIL -- thanks!
